src/backend/udev.rs

```diff
@@ -0,0 +1,402 @@
+                    },
+                    // Device removed
+                    EventType::Remove => {
+                        evlh.state().with_value(token, |state, udev| {
+                            info!(udev.logger, "Device Remove");
+                            if let Some(devnum) = event.devnum() {
+                                if let Some((device, fd_event_source)) = udev.devices.remove(&devnum) {
+                                    fd_event_source.remove();
+                                    let mut state: StateProxy = state.into();
+                                    udev.handler.device_removed(&mut state, &device);
+                                    let device = state.remove(device);
+                                    let fd = device.as_raw_fd();
+                                    drop(device);
+                                    if let Err(err) = udev.session.close(fd) {
+                                        warn!(udev.logger, "Failed to close device {:?}. Error: {:?}. Ignoring", event.sysname(), err);
+                                    };
+                                }
+                            }
+                        })
+                    },
+                    // New connector
+                    EventType::Change => evlh.state().with_value(token, |state, udev| {
+                        info!(udev.logger, "Device Changed");
+                        if let Some(devnum) = event.devnum() {
+                            info!(udev.logger, "Devnum: {:b}", devnum);
+                            if let Some(&(ref device, _)) = udev.devices.get(&devnum) {
+                                info!(udev.logger, "changed successful");
+                                udev.handler.device_changed(state, device);
+                            } else {
+                                info!(udev.logger, "changed, but device not tracked by backend");
+                            }
+                        } else {
+                            info!(udev.logger, "changed, but no devnum");
+                        }
+                    }),
+                    _ => {},
+                }
+            }
+        },
+        error: |evlh, token, _, err| {
+            evlh.state().with_value(token, |state, udev| udev.handler.error(state, err))
+        },
+    }
+}
+
+/// Handler for the `UdevBackend`, allows to open, close and update drm devices as they change during runtime.
+pub trait UdevHandler<B: Borrow<DrmBackend> + 'static, H: DrmHandler<B> + 'static> {
+    /// Called on initialization for every known device and when a new device is detected.
+    ///
+    /// Returning a `DrmHandler` will initialize the device, returning `None` will ignore the device.
+    ///
+    /// ## Panics
+    /// Panics if you try to borrow the token of the belonging `UdevBackend` using this `StateProxy`.
+    fn device_added<'a, S: Into<StateProxy<'a>>>(&mut self, state: S, device: &mut DrmDevice<B>) -> Option<H>;
```

Comment:
If I understand correctly, this `S: Into<StateProxy>` is needed because the `UdevHandler` instance will be stored in a value that will itself be stored in the EventLoop's State.

Is storing this into the EventLoop's state absolutely necessary, or could using the implementation data be an option too? This would make it possible to not have to borrow the state to retrieve the handler, and as such probably allow to directly provide a `&mut EventLoopHandle`, I think?

Reply:
It is necessary to register the `UdevBackend` to the `SessionNotifier`, which is also the reason, why I have a separate `bind`-function.

I have already evaluated, if I could remove any bind calls and add them to the constructor, but for all remaining values I did not find a way to do that with the current api.

`examples/udev.rs` really shows what needs to be done with all values at the minimum, if you have concerns about the general api or about any specific case fine, but the overall structure of `new` -> `register`/user-calls -> `bind` is kinda necessary.

Also we were talking about building a extensible and composable library. I would rather not want to force the user to use the `State` or the `EventLoop` at all, where easily possible.